txp/Pos/Txp/Logic/Local.hs

```diff
@@ -79,68 +74,48 @@ txProcessTransaction itw@(txId, txAux) = do
   where
-    notBootRelated bootHolders =
-        let txDistr = getTxDistribution $ taDistribution txAux
-            inBoot s = s `HS.member` bootHolders
-            mentioned pool addr = addr `elem` pool
-            bad :: [(StakeholderId, Coin)] -> Bool
-            bad (map fst -> txOutDistr) =
-                -- Has unrelated address
-                any (not . inBoot) txOutDistr ||
-                -- Not all genesis boot addrs are mentioned
-                any (not . mentioned txOutDistr) (HS.toList bootHolders)
-        in NE.filter bad txDistr
-
     processTxDo
-        :: Utxo
-        -> HashSet StakeholderId
-        -> ToilEnv
+        :: EpochIndex
+        -> BlockVersionData
+        -> GenesisStakeholders
+        -> Utxo
         -> HeaderHash
         -> (TxId, TxAux)
-        -> Bool
         -> TxpLocalDataPure
         -> (Either ToilVerFailure (), TxpLocalDataPure)
-    processTxDo resolved bootHolders toilEnv tipDB tx bootEra txld@(uv, mp, undo, tip, ()) = do
-        let tipMismatch = tipDB /= tip
-            bootRel = notBootRelated bootHolders
-        if | tipMismatch ->
-             (Left $ ToilTipsMismatch tipDB tip, txld)
-           | bootEra && not (null bootRel) ->
-            -- We do not allow txs with non-boot-addr stake distr in boot era.
-             (Left $ ToilBootDifferentStake $ unsafeHead bootRel, txld)
-           | otherwise ->
-             let res = (runExceptT $
-                        flip evalUtxoStateT resolved $
-                        execToilTLocal uv mp undo $
-                        processTx tx) toilEnv
-             in case res of
-                 Left er  -> (Left er, txld)
-                 Right ToilModifier{..} ->
-                     (Right (), (_tmUtxo, _tmMemPool, _tmUndos, tip, ()))
+    processTxDo curEpoch bvd genStks resolved tipDB tx txld@(uv, mp, undo, tip, ())
+        | tipDB /= tip =
+            (Left $ ToilTipsMismatch tipDB tip, txld)
+        | otherwise =
+            let res = (runExceptT $
+                    flip runReaderT genStks $
+                    flip evalUtxoStateT resolved $
+                    execToilTLocal uv mp undo $
+                    processTx curEpoch tx) bvd
+            in case res of
+                Left er  -> (Left er, txld)
+                Right ToilModifier{..} ->
+                    (Right (), (_tmUtxo, _tmMemPool, _tmUndos, tip, ()))
 
 -- | 1. Recompute UtxoView by current MemPool
 -- | 2. Remove invalid transactions from MemPool
 -- | 3. Set new tip to txp local data
-txNormalize ::
-       ( MonadIO m
-       , MonadBaseControl IO m
-       , MonadDBRead m
-       , MonadGState m
-       , MonadTxpMem () ctx m
-       , WithLogger m
-       )
+txNormalize
+    :: ( TxpLocalWorkMode ctx m
+       , MonadSlots m)
     => m ()
 txNormalize = do
     utxoTip <- GS.getTip
+    epoch <- maybe (throwM ToilUnknownCurEpoch) (pure . siEpoch) =<< getCurrentSlot
```

Comment:
No, it's broken. When we load blocks after long offline, we don't know the current slot. This function will throw and something will go wrong.
The proper solution is to just clean mempool if we don't know the current slot.